vllm/model_executor/models/transformers.py

```diff
@@ -204,36 +300,55 @@ def apply_base_model_tp_plan(self, module: nn.Module, prefix: str = ""):
             else:
                 self.apply_base_model_tp_plan(child_module, prefix=qual_name)
 
-    def replace_vocab_embed_class(self, module: nn.Module):
-        # Use native set input embeddings
-        new_module = VocabParallelEmbedding(
-            self.vocab_size,
-            self.config.hidden_size,
-            org_num_embeddings=self.config.vocab_size,
-            quant_config=None,
-        )
-        log_replacement("input embedding", self.model.get_input_embeddings(),
-                        new_module)
-        self.model.set_input_embeddings(new_module)
+    def init_buffers(self, module: nn.Module):
+        for name, buffer in module.named_buffers(recurse=False):
+            if buffer.device == torch.device("meta"):
+                new_buffer = getattr(type(module)(self.config), name)
+                setattr(module, name, new_buffer)
+        for child in module.children():
+            self.init_buffers(child)
+
+    def meta_to_empty(self, module: nn.Module):
+        tensors = list(chain(module.buffers(), module.parameters()))
+        if tensors and all(t.device == torch.device("meta") for t in tensors):
+            module.to_empty(device=self.device_config.device)
+            return  # We can stop recursing because to_empty is recursive
+        for child in module.children():
+            self.meta_to_empty(child)
 
     def forward(
         self,
-        input_ids: torch.Tensor,
+        input_ids: Optional[torch.Tensor],
         positions: torch.Tensor,
         kv_caches: list[torch.Tensor],  # argument not used
         attn_metadata: AttentionMetadata,
         intermediate_tensors: Optional[IntermediateTensors] = None,
         inputs_embeds: Optional[torch.Tensor] = None,
     ) -> Union[torch.Tensor, IntermediateTensors]:
-        model_output = self.model(
-            input_ids[None, ...],
+        if get_pp_group().is_first_rank:
+            if input_ids is not None:
+                input_ids = input_ids[None, ...]
+            if inputs_embeds is not None:
+                inputs_embeds = inputs_embeds[None, ...]
```

Comment:
Why is this logic necessary in only the PP case?

Reply:
You're right, I'll move this